### PR TITLE
Fix EncoderASR interface that was very restrictive 

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

This PR fix the `EncoderASR` class which was forcing the users to add a new parameter in our HF inference YAMLs `decoding_type`. Furthermore, this parameter was not letting the possibility to the end user to customise easily the type of beam search used, the hyperparams etc. 

Now, we do not need this parameter. We can automatically retrieve if it's greedy search or beam search. You can either use greedy search like that:
```yaml
decoding_function: !name:speechbrain.decoders.ctc_greedy_decode
    blank_id: !ref <blank_index>
```
Or beam search:
```yaml
decoding_function: !name:speechbrain.decoders.CTCBeamSearcher
# Decoding parameters
test_beam_search:
   beam_size: 143
   topk: 1
   blank_index: !ref <blank_index>
   space_token: ' ' # make sure this is the same as the one used in the tokenizer
   beam_prune_logp: -12.0
   token_prune_min_logp: -1.2
   prune_history: True
   alpha: 0.8
   beta: 1.2
   # can be downloaded from here https://www.openslr.org/11/ or trained with kenLM
   # It can either be a .bin or .arpa ; note: .arpa is much slower at loading
   # If you don't want to use an LM, comment it out or set it to null
   kenlm_model_path: null
```

Furthermore, for a mysterious reason, the class was expecting Sentencepiece tokenizer while most of our models expect CTCTextEncoder... So this PR is also solving this issue by checking the tokenizer before retrieving the vocab_list. 

Finally, this PR adds a small check that the input log_probs are effectively log_probs. I found that in wav2vec2 HF models, we were not passing the log softmax in the encoder which means that the logits (and not log probs) were given to the beam search decoder and ultimately led to wrong results!

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

**Comment by Adel-Moumen** [comment]
I tried it and got the expected output on all decoders (greedy, beam search, beam search + ngram):
CTC model greedy - THE BIRCH CANOE SLID ON THE SMOOTH PLANKS
CTC model beam search - THE BIRCH CANOE SLID ON THE SMOOTH PLANKS
CTC model beam search + ngram - THE BIRCH CANOE SLID ON THE SMOOTH PLANKS

To reproduce:
```bash
wget https://openslr.elda.org/resources/11/3-gram.pruned.3e-7.arpa.gz
gzip -d 3-gram.pruned.3e-7.arpa.gz
```

And then, modify the YAML so that you add the path to the n-gram.

File: speechbrain/decoders/ctc.py [hunk 1]
```diff
@@ -965,6 +965,16 @@ def decode_beams(
                 "During decoding, going to truncate the log_probs vocab dim to match vocab_list."
             )
 
+        # check if we have log_probs
+        if not torch.allclose(
```
Comment (TParcollet) [inline speechbrain/decoders/ctc.py:969]: This operation is called for every call, this might be a problem for people that have speed requirements.

Comment (Adel-Moumen) [reply speechbrain/decoders/ctc.py:969]: Yes I dunno. I can remove it if you think it is up to the user to make sure that log probs are passed in input, otherwise, I can add an input flag that is by default True, and can be switch to False if one wants to remove this check. wdyt ?

Comment (TParcollet) [reply speechbrain/decoders/ctc.py:969]: Yes remove it, and add info in the doc about the requirements of the input

**Review by TParcollet — APPROVED** [review]
LGTM
